pkg/apiserver/proxy_test.go

```diff
@@ -34,6 +36,115 @@ import (
+	// are different (i.e., if a Transfer-Encoding header field is present).
+	return int64(-1)
+}
+
+func containGzip(transferEncodings []string) bool {
+	for _, encoding := range transferEncodings {
+		if encoding == "gzip" {
+			return true
+		}
+	}
+	return false
+}
+
+func TestProxyRequestContentLengthAndTransferEncoding(t *testing.T) {
+	serverResponse := "got response"
+	table := []struct {
+		transferEncodings []string
+		reqBody           string
+		reqNamespace      string
+	}{
+		{[]string{}, "", "default"},
+		{[]string{"identity"}, "question", "default"},
+		{[]string{"chunked"}, "question", "default"},
+		// RFC2616 section-3.6: Whenever a transfer-coding is applied to a message-body, the set of
+		// transfer-codings MUST include "chunked", unless the message is terminated by closing the
+		// connection.
+		{[]string{"chunked", "gzip"}, "qqqqqqqqqquestion", "default"},
+	}
+
+	for _, item := range table {
+		proxyServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			expectedContentLength := calculateContentLength(item.reqBody, item.transferEncodings)
+			if e, a := expectedContentLength, req.ContentLength; e != a {
+				t.Errorf("expected %v, got %v", e, a)
+			}
+			// We expect Content-Length header field be set
+			if expectedContentLength != -1 {
+				contentLengthInHeader, err := strconv.Atoi(req.Header.Get("Content-Length"))
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
+				if e, a := expectedContentLength, int64(contentLengthInHeader); e != a {
+					t.Errorf("expected %v, got %v", e, a)
+				}
+			}
+
+			// Help wanted: why is the transfer-encoding header field always empty?
+			if e, a := "", req.Header.Get("Transfer-Encoding"); e != a {
+				t.Errorf("expected %v, got %v", e, a)
+			}
+			fmt.Fprint(w, serverResponse)
+		}))
+		defer proxyServer.Close()
+
+		serverURL, _ := url.Parse(proxyServer.URL)
+		simpleStorage := &SimpleRESTStorage{
+			errors:                    map[string]error{},
+			resourceLocation:          serverURL,
+			expectedResourceNamespace: item.reqNamespace,
+		}
+
+		namespaceHandler := handleNamespaced(map[string]rest.Storage{"foo": simpleStorage})
+		server := httptest.NewServer(namespaceHandler)
+		defer server.Close()
+
+		proxyTestPattern := "/api/version2/proxy/namespaces/default/foo/id/some/dir"
+		var reader io.Reader
+		if containGzip(item.transferEncodings) {
+			var b bytes.Buffer
+			gzw := gzip.NewWriter(&b)
+			if _, err := gzw.Write([]byte(item.reqBody)); err != nil {
+				t.Errorf("unexpected error: %v", err)
+			}
+			reader = &b
+		} else {
+			reader = strings.NewReader(item.reqBody)
+		}
+		req, err := http.NewRequest(
+			"POST",
+			server.URL+proxyTestPattern,
+			reader,
+		)
+		if err != nil {
+			t.Errorf("unexpected error %v", err)
+			continue
+		}
+		req.TransferEncoding = item.transferEncodings
+		req.Header.Set("Transfer-Encoding", strings.Join(item.transferEncodings, ","))
```

Comment:
I think you should call Header.Add for each one instead of joining them manually. https://golang.org/pkg/net/http/#Header.Add